lighthouse-core/test/gather/driver-test.js

```diff
@@ -421,27 +509,133 @@ describe('.gotoURL', () => {
     expect(loadedUrl).toEqual(finalUrl);
   });
 
-  describe('when waitForNavigated', () => {});
+  describe('when waitForNavigated', () => {
+    it('waits for Page.frameNavigated', async () => {
+      driver.on = driver.once = createMockOnceFn()
+        .mockEvent('Page.frameNavigated');
+
+      const url = 'https://www.example.com';
+      const loadOptions = {
+        waitForNavigated: true,
+      };
+
+      const loadPromise = driver.gotoURL(url, loadOptions);
+      await flushAllTimersAndMicrotasks();
+      await loadPromise;
+    });
+  });
 
   describe('when waitForLoad', () => {
+    const url = 'https://example.com';
+
+    ['FCP', 'LoadEvent', 'NetworkIdle', 'CPUIdle'].forEach(name => {
+      it(`should wait for ${name}`, async () => {
+        driver._waitForFCP = createMockListenerFn();
+        driver._waitForLoadEvent = createMockListenerFn();
+        driver._waitForNetworkIdle = createMockListenerFn();
+        driver._waitForCPUIdle = createMockListenerFn();
+
+        const listener = driver[`_waitFor${name}`];
+        const otherListeners = [
+          driver._waitForFCP,
+          driver._waitForLoadEvent,
+          driver._waitForNetworkIdle,
+          driver._waitForCPUIdle,
+        ].filter(l => l !== listener);
```

Comment:
just learned that computed property names work with destructuring

```js
const waitFns = {
  _waitForFCP: createMockListenerFn(),
  _waitForLoadEvent: createMockListenerFn(),
  _waitForNetworkIdle: createMockListenerFn(),
  _waitForCPUIdle: createMockListenerFn(),
};
const {[`_waitFor${name}`]: listener, ...otherListeners} = waitFns;

Object.assign(driver, waitFns);
```

But I can't decide how evil that is :)

Reply:
😮